sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfiguration/provider/_azureappconfigurationprovider.py

```diff
@@ -352,3 +472,47 @@ def __exit__(self, *args) -> None:
         self._client.__exit__(*args)
         for client in self._secret_clients.values():
             client.__exit__()
+
+    class _ConfigurationRefresh:
+        def __init__(self, **kwargs):
+            self.refresh_options: Optional[AzureAppConfigurationRefreshOptions] = kwargs.pop(
+                "refresh_options", AzureAppConfigurationRefreshOptions()
+            )
+            if self.refresh_options is not None:
+                self.next_refresh_time = datetime.now() + timedelta(seconds=self.refresh_options.refresh_interval)
+            self.attempts = 1
+            self.min_backoff: Optional[int] = kwargs.pop("min_backoff", 30)
+            self.max_backoff: Optional[int] = kwargs.pop("min_backoff", 600)
+
+        def updated_configurations(self):
+            # pylint:disable=protected-access
+            self.next_refresh_time = datetime.now() + timedelta(seconds=self.refresh_options.refresh_interval)
+            self.attempts = 1
+            self.refresh_options._callback()
+
+        def failed_update(self, error, message):
+            logging.warning(message, self.attempts, error)
+            # Refresh All or None, any failure will trigger a backoff
+            # pylint:disable=protected-access
+            self.next_refresh_time = datetime.now() + timedelta(microseconds=self.calculate_backoff())
+            self.attempts += 1
+            self.refresh_options._on_error()
+
+        def calculate_backoff(self):
+            max_attempts = 30
+            microsecond = 1000000  # 1 Second in microseconds
+
+            min_backoff_microseconds = self.min_backoff * microsecond
+            max_backoff_microseconds = self.max_backoff * microsecond
+
+            if self.attempts <= 1 or self.max_backoff <= self.min_backoff:
```

Comment:
We also need to take into account the refresh interval of this sentinel key. The max_backoff is actually the refresh interval because we cannot refresh any later than the specified refresh interval.
See the .NET implementation here: https://github.com/Azure/AppConfiguration-DotnetProvider/blob/9a76d2a1b43d137e139e329a575eee09641e51d3/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/Extensions/TimeSpanExtensions.cs#L30-L37

Reply:
> We also need to take into account the refresh interval of this sentinel key.

We just have one refresh interval, not one per sentinel key?

Reply:
Yes, what I meant was we need to calculate backoff based on the refresh_interval set by users. Max backoff cannot be 10 minutes if the refresh_interval is 1 minute. We need to refresh at least every 1 minute if the refresh_interval is 1 minute.